### PR TITLE
Update ua_binary.py and Node.py

### PR DESCRIPTION
**Pull request by AndreasHeine**
fix: weird bug -> Server_GetMonitoredItems
fix: str/repr in node.py

### CODE REVIEW

**Comment by AndreasHeine** [comment]
@oroulet :
> where doeas the ns=0 nodes came from spec files? the arraydimensions 0 must came from somewhere!

can you tell me or show me the xml file (link)? found something maybe we dont need to change code!

https://github.com/FreeOpcUa/opcua-asyncio/blob/master/asyncua/server/standard_address_space/standard_address_space_part5.py

```
    node = ua.AddNodesItem()
    node.RequestedNewNodeId = NumericNodeId(11492, 0)
    node.BrowseName = QualifiedName('GetMonitoredItems', 0)
    node.NodeClass = NodeClass.Method
    node.ParentNodeId = NumericNodeId(2253, 0)
    node.ReferenceTypeId = NumericNodeId(47, 0)
    attrs = ua.MethodAttributes()
    attrs.DisplayName = LocalizedText("GetMonitoredItems")
    node.NodeAttributes = attrs
    server.add_nodes([node])
    refs = []
    ref = ua.AddReferencesItem()
    ref.IsForward = True
    ref.ReferenceTypeId = NumericNodeId(46, 0)
    ref.SourceNodeId = NumericNodeId(11492, 0)
    ref.TargetNodeClass = NodeClass.DataType
    ref.TargetNodeId = NumericNodeId(11493, 0)
    refs.append(ref)
    ref = ua.AddReferencesItem()
    ref.IsForward = True
    ref.ReferenceTypeId = NumericNodeId(46, 0)
    ref.SourceNodeId = NumericNodeId(11492, 0)
    ref.TargetNodeClass = NodeClass.DataType
    ref.TargetNodeId = NumericNodeId(11494, 0)
    refs.append(ref)
    ref = ua.AddReferencesItem()
    ref.IsForward = False
    ref.ReferenceTypeId = NumericNodeId(47, 0)
    ref.SourceNodeId = NumericNodeId(11492, 0)
    ref.TargetNodeClass = NodeClass.DataType
    ref.TargetNodeId = NumericNodeId(2253, 0)
    refs.append(ref)
    server.add_references(refs)
    node = ua.AddNodesItem()
    node.RequestedNewNodeId = NumericNodeId(11493, 0)
    node.BrowseName = QualifiedName('InputArguments', 0)
    node.NodeClass = NodeClass.Variable
    node.ParentNodeId = NumericNodeId(11492, 0)
    node.ReferenceTypeId = NumericNodeId(46, 0)
    node.TypeDefinition = NumericNodeId(68, 0)
    attrs = ua.VariableAttributes()
    attrs.DisplayName = LocalizedText("InputArguments")
    attrs.DataType = NumericNodeId(296, 0)
    value = []
    extobj = ua.Argument()
    extobj.Name = 'SubscriptionId'
    extobj.DataType = NumericNodeId(7, 0)
    extobj.ValueRank = -1
    value.append(extobj)
    attrs.Value = ua.Variant(value, ua.VariantType.ExtensionObject)
    attrs.ValueRank = 1
    attrs.ArrayDimensions = [0]
    node.NodeAttributes = attrs
    server.add_nodes([node])
    refs = []
    ref = ua.AddReferencesItem()
    ref.IsForward = True
    ref.ReferenceTypeId = NumericNodeId(40, 0)
    ref.SourceNodeId = NumericNodeId(11493, 0)
    ref.TargetNodeClass = NodeClass.DataType
    ref.TargetNodeId = NumericNodeId(68, 0)
    refs.append(ref)
    ref = ua.AddReferencesItem()
    ref.IsForward = False
    ref.ReferenceTypeId = NumericNodeId(46, 0)
    ref.SourceNodeId = NumericNodeId(11493, 0)
    ref.TargetNodeClass = NodeClass.DataType
    ref.TargetNodeId = NumericNodeId(11492, 0)
    refs.append(ref)
    server.add_references(refs)
   
    node = ua.AddNodesItem()
    node.RequestedNewNodeId = NumericNodeId(11494, 0)
    node.BrowseName = QualifiedName('OutputArguments', 0)
    node.NodeClass = NodeClass.Variable
    node.ParentNodeId = NumericNodeId(11492, 0)
    node.ReferenceTypeId = NumericNodeId(46, 0)
    node.TypeDefinition = NumericNodeId(68, 0)
    attrs = ua.VariableAttributes()
    attrs.DisplayName = LocalizedText("OutputArguments")
    attrs.DataType = NumericNodeId(296, 0)
    value = []
    extobj = ua.Argument()
    extobj.Name = 'ServerHandles'
    extobj.DataType = NumericNodeId(7, 0)
    extobj.ValueRank = 1
    extobj.ArrayDimensions = 0 <-----------------------------------------------------------!
    value.append(extobj)
    extobj = ua.Argument()
    extobj.Name = 'ClientHandles'
    extobj.DataType = NumericNodeId(7, 0)
    extobj.ValueRank = 1
    extobj.ArrayDimensions = 0 <-----------------------------------------------------------!
    value.append(extobj)
    attrs.Value = ua.Variant(value, ua.VariantType.ExtensionObject)
    attrs.ValueRank = 1
    attrs.ArrayDimensions = [0]
    node.NodeAttributes = attrs
    server.add_nodes([node])
    refs = []
    ref = ua.AddReferencesItem()
    ref.IsForward = True
    ref.ReferenceTypeId = NumericNodeId(40, 0)
    ref.SourceNodeId = NumericNodeId(11494, 0)
    ref.TargetNodeClass = NodeClass.DataType
    ref.TargetNodeId = NumericNodeId(68, 0)
    refs.append(ref)
    ref = ua.AddReferencesItem()
    ref.IsForward = False
    ref.ReferenceTypeId = NumericNodeId(46, 0)
    ref.SourceNodeId = NumericNodeId(11494, 0)
    ref.TargetNodeClass = NodeClass.DataType
    ref.TargetNodeId = NumericNodeId(11492, 0)
    refs.append(ref)
    server.add_references(refs)
   
```

**Comment by oroulet** [comment]
Most probably the arrayDimensions to 0 is either an xml parsing bug or some strange xml spec from opc foundation

**Comment by oroulet** [comment]
The code over looks ok. Doesn't it?

**Comment by AndreasHeine** [comment]
the code change would auto correct it but maybe its better to check parsing!? dont know how many changes will come soon

**Comment by oroulet** [comment]
What xml do you want? The one in schemas directory?

**Comment by AndreasHeine** [comment]
> The one in schemas directory?

i feel silly and blind... where?

```
"""
DO NOT EDIT THIS FILE!
It is automatically generated from opcfoundation.org schemas.
"""
```

**Comment by oroulet** [comment]
The python file is called generate_addressspace.py in schemas directory in repository

**Comment by oroulet** [comment]
Ok. Maybe the xml has been removed

**Comment by oroulet** [comment]
the xml is there.origin	https://github.com/OPCFoundation/UA-Nodeset/ (fetch)
origin	https://github.com/OPCFoundation/UA-Nodeset/ (push)

just clone it schemas directory. Ithought this was documented but maybe I forgot.. We should add a README in that directory to explain

**Comment by AndreasHeine** [comment]
i will have a look... maybe i find something!

**Comment by oroulet** [comment]
the code is there:
https://github.com/FreeOpcUa/opcua-asyncio/blob/master/schemas/generate_address_space.py

I see a bunch of things are hard coded for array dimension. it might be easy to fix

**Comment by oroulet** [comment]
and please add a README in that folder to explain how to run things if there is nothing

**Comment by AndreasHeine** [comment]
> and please add a README in that folder to explain how to run things if there is nothing

sure!

**Comment by AndreasHeine** [comment]
```
  <UAVariable NodeId="i=11493" BrowseName="InputArguments" ParentNodeId="i=11492" DataType="i=296" ValueRank="1" ArrayDimensions="0">
    <DisplayName>InputArguments</DisplayName>
    <References>
      <Reference ReferenceType="HasTypeDefinition">i=68</Reference>
      <Reference ReferenceType="HasProperty" IsForward="false">i=11492</Reference>
    </References>
    <Value>
      <ListOfExtensionObject xmlns="http://opcfoundation.org/UA/2008/02/Types.xsd">
        <ExtensionObject>
          <TypeId>
            <Identifier>i=297</Identifier>
          </TypeId>
          <Body>
            <Argument>
              <Name>SubscriptionId</Name>
              <DataType>
                <Identifier>i=7</Identifier>
              </DataType>
              <ValueRank>-1</ValueRank>
              <ArrayDimensions />
            </Argument>
          </Body>
        </ExtensionObject>
      </ListOfExtensionObject>
    </Value>
  </UAVariable>
```
<ValueRank>-1</ValueRank> => `n = -1: the value is a scalar`

```
<UAVariable NodeId="i=11494" BrowseName="OutputArguments" ParentNodeId="i=11492" DataType="i=296" ValueRank="1" ArrayDimensions="0">
    <DisplayName>OutputArguments</DisplayName>
    <References>
      <Reference ReferenceType="HasTypeDefinition">i=68</Reference>
      <Reference ReferenceType="HasProperty" IsForward="false">i=11492</Reference>
    </References>
    <Value>
      <ListOfExtensionObject xmlns="http://opcfoundation.org/UA/2008/02/Types.xsd">
        <ExtensionObject>
          <TypeId>
            <Identifier>i=297</Identifier>
          </TypeId>
          <Body>
            <Argument>
              <Name>ServerHandles</Name>
              <DataType>
                <Identifier>i=7</Identifier>
              </DataType>
              <ValueRank>1</ValueRank>
              <ArrayDimensions>
                <UInt32>0</UInt32>
              </ArrayDimensions>
            </Argument>
          </Body>
        </ExtensionObject>
        <ExtensionObject>
          <TypeId>
            <Identifier>i=297</Identifier>
          </TypeId>
          <Body>
            <Argument>
              <Name>ClientHandles</Name>
              <DataType>
                <Identifier>i=7</Identifier>
              </DataType>
              <ValueRank>1</ValueRank>
              <ArrayDimensions>
                <UInt32>0</UInt32>
              </ArrayDimensions>
            </Argument>
          </Body>
        </ExtensionObject>
      </ListOfExtensionObject>
    </Value>
  </UAVariable>
```

<ValueRank>1</ValueRank> => `n >= 1: the value is an array with the specified number of dimensions`

looks right on the first look.... now generate_address_space.py

**Comment by AndreasHeine** [comment]
>         if obj.dimensions:
>             self.writecode(indent, f'attrs.ArrayDimensions = {obj.dimensions}')
> 
> that code is a bit suspicious. youcan probably make sure obj.dimensions is a list and if not maybe just write None.Was does0 means btw? I suspect writing None is correct, or?

hm guess we need to check ValueRank before we write ArrayDimensions to we now exacly how many we've got!

> Value Rank
> 
> This attribute indicates whether the value attribute of the variable is an array and how many dimensions the array has. It may have the following values:
> 
>     n >= 1: the value is an array with the specified number of dimensions
>     n =  0: the value is an array with one or more dimensions
>     n = -1: the value is a scalar
>     n = -2: the value can be a scalar or an array with any number of dimensions
>     n = -3: the value can be a scalar or a one dimensional array
> 
> Consistency between the value rank attribute in the variable and its VariableTypeNode is ensured.
> Array Dimensions
> 
> If the value rank permits the value to be a (multi-dimensional) array, the exact length in each dimensions can be further constrained with this attribute.
> 
>     For positive lengths, the variable value is guaranteed to be of the same length in this dimension.
>     The dimension length zero is a wildcard and the actual value may have any length in this dimension.
> 
> Consistency between the array dimensions attribute in the variable and its VariableTypeNode is ensured.

anyways i will have a look... ;)

**Comment by AndreasHeine** [comment]
```
    def make_ext_obj_code(self, indent, extobj):
        self.writecode(indent, f'extobj = ua.{extobj.objname}()')
        for name, val in extobj.body:
            for k, v in val:
                if type(v) is str:
                    val = _to_val([extobj.objname], k, v)
                    self.writecode(indent, f'extobj.{k} = {val}')
                else:
                    if k == "DataType":  #hack for strange nodeid xml format
                        self.writecode(indent, 'extobj.{0} = {1}'.format(k, nodeid_code(v[0][1])))
                        continue
                    if k == "ArrayDimensions":  # hack for v1.04 - Ignore UInt32 tag?
                        self.writecode(indent, 'extobj.ArrayDimensions = {0}'.format(v[0][1]))
                        continue
                    for k2, v2 in v:
                        val2 = _to_val([extobj.objname, k], k2, v2)
                        self.writecode(indent, f'extobj.{k}.{k2} = {val2}')
```

Solution:
`self.writecode(indent, 'extobj.ArrayDimensions = [{0}]'.format(v[0][1]))`

like always something small ;) just brackets

@oroulet : i'll make a new PR for this! with generation-readme and i will change some things in the path thing to make it usable for linux and windows users 
like: 
```
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
xml_path = f'{BASE_DIR}/schemas/UA-Nodeset/Schema/Opc.Ua.NodeSet2.Part{i}.xml'
py_path = f'{BASE_DIR}/asyncua/server/standard_address_space/standard_address_space_part{i}.py'
```

File: asyncua/common/node.py [hunk 1]
```diff
@@ -70,10 +70,12 @@ def __ne__(self, other):
         return not self.__eq__(other)
 
     def __str__(self):
+        node_str = str(self.nodeid)
+        return node_str[node_str.find("(")+1:-1]
```
Comment (oroulet) [inline asyncua/common/node.py:74]: node_str= node_str.replace("\"", "")   
Isn'it much simpler?

Comment (oroulet) [reply asyncua/common/node.py:74]: also the entire stuff is strange. why do you get quotes at all? where do they come from?

Comment (oroulet) [reply asyncua/common/node.py:74]: `return self.nodeid.to_string()`should be enough

Comment (AndreasHeine) [reply asyncua/common/node.py:74]: havn't seen that, of course its better!

**Comment by AndreasHeine** [comment]
@oroulet : https://github.com/FreeOpcUa/opcua-asyncio/compare/master...AndreasHeine:master
still comparing...  i generated all new, there are some changes guess from uanodesets but still looking!

**Pull request closed without merging**